typed-racket-lib/typed-racket/rep/type-rep.rkt

```diff
@@ -611,191 +788,136 @@
-  [#:intern (map Rep-seq tys)]
-  [#:frees (λ (f) (combine-frees (map f tys)))]
-  [#:key #f] [#:fold-rhs (*Sequence (map type-rec-id tys))])
-
-(def-type Future ([t Type/c]) [#:key 'future])
-
-;; body: the type of the body
-;; handler: the type of the prompt handler
-;;   prompts with this tag will return a union of `body` 
-;;   and the codomains of `handler`
-(def-type Prompt-Tagof ([body Type/c] [handler Type/c])
-  [#:frees (λ (f) (combine-frees (list (make-invariant (f body))
-                                       (make-invariant (f handler)))))]
-  [#:key 'prompt-tag])
-
-;; value: the type of allowable values
-(def-type Continuation-Mark-Keyof ([value Type/c])
-  [#:frees (λ (f) (make-invariant (f value)))]
-  [#:key 'continuation-mark-key])
+(def-type Sequence ([tys (listof Type?)])
+  [#:intern-key (map Rep-seq tys)]
+  [#:frees (f) (combine-frees (map f tys))]
+  [#:fold (f) (make-Sequence (map f tys))]
+  [#:walk (f) (for-each f tys)])
 
 ;; Distinction
 ;; comes from define-new-subtype
 ;; nm: a symbol representing the name of the type
 ;; id: a symbol created with gensym
-;; ty: a type for the representation, where this will be a subtype of ty
-(def-type Distinction ([nm symbol?] [id symbol?] [ty Type/c])
-  [#:frees (λ (f) (f ty))]
-  [#:intern (list (Rep-seq ty) nm id)]
-  [#:fold-rhs (*Distinction nm id (type-rec-id ty))]
-  [#:key (Type-key ty)])
-
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-
-;; remove-dups: List[Type] -> List[Type]
-;; removes duplicate types from a SORTED list
-(define/cond-contract (remove-dups types)
-  ((listof Rep?) . -> . (listof Rep?))
-  (cond [(null? types) types]
-        [(null? (cdr types)) types]
-        [(type-equal? (car types) (cadr types)) (remove-dups (cdr types))]
-        [else (cons (car types) (remove-dups (cdr types)))]))
+;; ty: a type for the representation (i.e. each distinction
+;;     is a subtype of its ty)
+(def-type Distinction ([nm symbol?] [id symbol?] [ty Type?])
+  [#:intern-key (list* nm id (Rep-seq ty))]
+  [#:frees (f) (f ty)]
+  [#:fold (f) (make-Distinction nm id (f ty))]
+  [#:walk (f) (f ty)]
+  [#:type-mask (Type-mask ty)])
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
 
-(define ((sub-f st) e)
-  (prop-case (#:Type st
-              #:Prop (sub-f st)
-              #:PathElem (sub-pe st))
-             e))
-
-
-(define ((sub-o st) e)
-  (object-case (#:Type st
-                #:Object (sub-o st)
-                #:PathElem (sub-pe st))
-               e))
-
-(define ((sub-pe st) e)
-  (pathelem-case (#:Type st
-                         #:PathElem (sub-pe st))
-                 e))
-
-(define ((sub-t st) e)
-  (type-case (#:Type st
-              #:Prop (sub-f st))
-              e))
-
-
-;; abstract-many : Names Type -> Type
-;; where n is the length of names
-(define (abstract-many names ty)
-  ;; mapping : dict[Type -> Natural]
-  (define (nameTo mapping type)
-    (let loop ([outer 0] [ty type])
```

Comment:
I think we just need to make this a 'generic' operation (since it gets called so frequently and that would cut out several sources of overhead)